python/cog/server/clients.py

```diff
@@ -0,0 +1,319 @@
+    # backoff. In total it'll try for up to roughly 320 seconds, providing
+    # resilience through temporary networking and availability issues.
+    transport = RetryTransport(
+        max_attempts=12,
+        backoff_factor=0.1,
+        retry_status_codes=[429, 500, 502, 503, 504],
+        retryable_methods=["POST"],
+    )
+    return httpx.AsyncClient(
+        event_hooks={"request": [on_request_trace_context_hook]},
+        headers=webhook_headers(),
+        transport=transport,
+        follow_redirects=True,
+    )
+
+
+def httpx_file_client() -> httpx.AsyncClient:
+    # verify: Union[str, bool, ssl.SSLContext] = True
+    transport = RetryTransport(
+        max_attempts=3,
+        backoff_factor=0.1,
+        retry_status_codes=[408, 429, 500, 502, 503, 504],
+        retryable_methods=["PUT"],
+        verify=os.environ.get("CURL_CA_BUNDLE", True),
+    )
+    # set connect timeout to slightly more than a multiple of 3 to avoid
+    # aligning perfectly with TCP retransmission timer
+    # requests has no write timeout, keep that
+    # httpx default for pool is 5, use that
+    timeout = httpx.Timeout(connect=10, read=15, write=None, pool=5)
+
+    return httpx.AsyncClient(
+        event_hooks={"request": [on_request_trace_context_hook]},
+        headers=common_headers(),
+        transport=transport,
+        follow_redirects=True,
+        timeout=timeout,
+        http2=True,
+    )
+
+
+class ChunkFileReader:
+    def __init__(self, fh: io.IOBase) -> None:
+        self.fh = fh
+
+    async def __aiter__(self) -> AsyncIterator[bytes]:
+        self.fh.seek(0)
+        while True:
+            chunk = self.fh.read(1024 * 1024)
+            if isinstance(chunk, str):
+                chunk = chunk.encode("utf-8")
+            if not chunk:
+                log.info("finished reading file")
+                break
+            yield chunk
+
+
+# there's a case for splitting this apart or inlining parts of it
+# I'm somewhat sympathetic to separating webhooks and files, but they both have
+# the same semantics of holding a client for the lifetime of runner
+# also, both are used by PredictionEventHandler
+
+
+class ClientManager:
+    def __init__(self) -> None:
+        self.webhook_client = httpx_webhook_client()
+        self.retry_webhook_client = httpx_retry_client()
+        self.file_client = httpx_file_client()
+        self.download_client = httpx.AsyncClient(follow_redirects=True, http2=True)
+
+    async def aclose(self) -> None:
+        # not used but it's not actually critical to close them
+        await self.webhook_client.aclose()
+        await self.retry_webhook_client.aclose()
+        await self.file_client.aclose()
+        await self.download_client.aclose()
+
+    # webhooks
+
+    async def send_webhook(
+        self, url: str, response: Dict[str, Any], event: WebhookEvent
+    ) -> None:
+        if Status.is_terminal(response["status"]):
```

Comment:
`event` is unused. Should we be consulting `event` instead of `response.status`? Or should we drop it from the method signature entirely?

Reply:
ah, right, this was moved up to make_webhook_sender